src/librustc_resolve/lib.rs

```diff
@@ -3148,49 +3148,64 @@ impl<'a, 'crateloader: 'a> Resolver<'a, 'crateloader> {
                         }
                         return (err, candidates);
                     },
-                    (Def::Struct(def_id), _) if ns == ValueNS => {
-                        if let Some((ctor_def, ctor_vis))
-                                = this.struct_constructors.get(&def_id).cloned() {
-                            let accessible_ctor = this.is_accessible(ctor_vis);
-                            if is_expected(ctor_def) && !accessible_ctor {
-                                err.span_label(span, format!("constructor is not visible \
-                                                              here due to private fields"));
-                            }
-                        } else {
-                            // HACK(estebank): find a better way to figure out that this was a
-                            // parser issue where a struct literal is being used on an expression
-                            // where a brace being opened means a block is being started. Look
-                            // ahead for the next text to see if `span` is followed by a `{`.
-                            let cm = this.session.source_map();
-                            let mut sp = span;
-                            loop {
-                                sp = cm.next_point(sp);
-                                match cm.span_to_snippet(sp) {
-                                    Ok(ref snippet) => {
-                                        if snippet.chars().any(|c| { !c.is_whitespace() }) {
-                                            break;
+                    (Def::Struct(def_id), PathSource::Expr(Some(parent))) if ns == ValueNS => {
```

Comment:
The failing UI tests that [Travis reports](https://github.com/rust-lang/rust/pull/54308#issuecomment-422201051) indicate that this code isn't behaving as we expected. What happens when the second element is some variant of `PathSource` other than `Expr`?